runtime/jsMain/src/kotlinx/serialization/DynamicObjectSerializer.kt

```diff
@@ -0,0 +1,167 @@
+import kotlinx.serialization.json.internal.END_LIST
+import kotlinx.serialization.json.internal.END_OBJ
+import kotlinx.serialization.modules.EmptyModule
+import kotlinx.serialization.modules.SerialModule
+import kotlinx.serialization.modules.getContextualOrDefault
+
+
+/**
+ * @param encodeDefaults if false we will always write nulls explicitly.
+ *          If true we will omit the property if it has a default value.
+ */
+class DynamicObjectSerializer(private val encodeDefaults: Boolean = false, val context: SerialModule = EmptyModule) {
+
+    fun <T> serialize(strategy: SerializationStrategy<T>, obj: T): dynamic {
+
+        // handle singular primitive and enum values immediately
+        if (strategy.descriptor.kind is PrimitiveKind) {
+            if (obj is Char) {
+                val value: Char = obj // without the explicit type annotation int value would be used
+                return value.toString()
+            }
+            if (obj is Long) {
+                return obj.toDouble()
+            }
+            return obj
+        }
+        if (strategy.descriptor.kind is UnionKind.ENUM_KIND) {
+            return obj.unsafeCast<Enum<*>>().name
+        }
+
+        val serializer = DynamicObjectEncoder(encodeDefaults)
+        serializer.encode(strategy, obj)
+        return serializer.result
+    }
+
+    @ImplicitReflectionSerializer
+    inline fun <reified T : Any> serialize(obj: T): dynamic = serialize(context.getContextualOrDefault(T::class), obj)
+
+    @ImplicitReflectionSerializer
+    inline fun <reified T : Any> serialize(obj: List<T?>): dynamic =
+        serialize(context.getContextualOrDefault(T::class).nullable.list, obj)
+}
+
+private class DynamicObjectEncoder(val encodeDefaults: Boolean) : ElementValueEncoder() {
+    private object NoOutputMark
+
+    var result: dynamic = NoOutputMark
+    lateinit var current: Node
+    var currentName: String = ""
+    var currentDescriptor: SerialDescriptor = StringDescriptor
+
+    class Node(val writeMode: WriteMode, val jsObject: dynamic) {
+        var index: Int = 0
+        lateinit var parent: Node
+    }
+
+    enum class WriteMode {
+        OBJ, MAP, LIST
+    }
+
+    override fun encodeElement(desc: SerialDescriptor, index: Int): Boolean {
+        current.index = index
+        currentDescriptor = desc
+
+        if (current.writeMode != WriteMode.MAP) {
+            currentName = desc.getElementName(index)
+        }
+        return true
+    }
+
+    override fun encodeValue(value: Any) {
+        val elementIsMapKey = current.writeMode == WriteMode.MAP && current.index % 2 == 0
+        if (elementIsMapKey) {
+            val elementDescriptor = currentDescriptor.getElementDescriptor(current.index)
+            currentName = if (elementDescriptor.kind is PrimitiveKind) {
+                value.toString()
+            } else if (elementDescriptor.kind is UnionKind.ENUM_KIND && value is Enum<*>) {
+                value.name
```

Comment:
Should we honor custom serializers or `@SerialName`/`@SerialId` annotations for enums when being used as map keys?

Reply:
We now have two cases where we (naively) serialize Enum as `Enum.name`. We will have to again resort to something like `DynamicValueEncoder` to do it correctly.

Also if we were supporting custom serialisers here, we would have to support them for all types  iff the serialiser returns a legal javascript object key (that is String or Int).

Reply:
> Also if we were supporting custom serialisers here, we would have to support them for all types iff the serialiser returns a legal javascript object key (that is String or Int).

Yeah that's fair. I think the most important thing is to be clear about what the limitations are.

Reply:
You can apply here the same technique that I described in previous comment

Reply:
However, there's a bigger problem with that: `encodeValue` is a fallback method, i.e. if map key is `Char` or `Enum`, `encodeChar` and `encodeEnum` are called, respectively — and this method won't be called. Seems like you have two options: 1. Move 'elementIsMapKey' check to a separate function and call it in the beginning of `encodeChar/Enum` 2. Move check to a `encodeElement` function (check itself does not require actual value, only descriptor) and then delegate to a `encodeValue` manually (e.g. `override fun encodeEnum(enumDescription: SerialDescriptor, ordinal: Int) = encodeValue(enumDescription.getElementName(ordinal))`

Reply:
Thanks for the suggestions, they helped to arrive a somewhat elegant solution.
We delegate writing operations to `encodeValue`. We also keep track of `currentElementIsMapKey` and use it in `beginStructure` to throw an exception and in  `encodeNull` to write null as the map key if necessary.

Reply:
We now support `@Serialname` on enums, both for map keys and for singular enums (and of course for regular properties).

Reply:
I also documented what is currently not supported which is 
```
 * Limitations:
 * * Map keys must be of primitive or enum type
 * * Custom serializers are not supported
 * * Enums are serialized as the value of `@SerialName` if present or their name, in that order. Ordinals are not supported.
 * * There is no analogue to DynamicObjectParsers' useArrayPolymorphism flag
```

Reply:
we now support custom serialisers everywhere :)